### ISSUE TITLE
REDUNDANCY_LEVEL is useless... :(

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#579]
Actually `REDUNDANCY_LEVEL = 2` is worse than useless: it causes `write file error` ... `index: 0 - error sent by master server (No chunk servers)` for replicated goal=3 when two chunkservers are stopped.

Goal is defined as `3: $std {stor pool wks}` and there are five (5) chunkservers labelled `wks`.  
Problem happens when I stop any two `wks` chunkservers while it should have been possible to stop all `wks` chunkservers...

**Comment by psarna** [comment]
Do you mean the situation where:
1. `REDUNDANCY_LEVEL = 2`
2. File goal is set to 3
3. Only 1 out of 3 copies are available, because 2 chunkservers that hold other copies are down
4. File is being **modified**?

If so, this is expected behaviour. If file data is to modified, then at least one of its chunks is going to be modified. Let's assume that we want to edit 1 byte in the middle of one of the chunks. With `REDUNDANCY_LEVEL = 2`, we want master to ensure that at least 2 copies of most recent data is always present. If master would allow clients to edit this 1 byte, then there would be only one copy of this most recent data, which is in conflict with `REDUNDANCY_LEVEL = 2`. Writing will be possible once at least 2 copies of the chunk are writable at the moment, which will happen after the only copy is replicated. That's how writing and replication works in LizardFS.
If, instead of a write error, you would expect the write operation to either "hang until other copies are replicated" or "read the chunk first and then write the modified version enough number of times", then there are caveats, races and performance choices to be considered, it's not a very trivial topic.

**Comment by onlyjob** [comment]
No, let me explain what's happening:

There are three groups of chunkservers labeled as _stor_, _pool_ and _wks_. Goal is defined as `3: $std {stor pool wks}` to put each replica to corresponding pool of chunkservers. There are 5 chunkservers in _wks_ pool. With `REDUNDANCY_LEVEL = 2` currently it is not possible to modify files when 2 chunkservers in _wks_ pool are stopped. This is certainly not expected as all chunkservers in pools _stor_ and _pool_ are available as well as 3 chunkservers in pool _wks_. There are enough operational chunkservers to update all three replicas and with this configuration it should be possible to update file even when all chunkservers in one pool are down.

**Comment by psarna** [comment]
Oh, I got it... this is an issue with `REDUNDANCY_LEVEL` documentation, not implementation. Here's how to interpret `REDUNDANCY_LEVEL = X`:
Each time master receives a write/modification request, it is going to check if at least original data + X **additional** parts are written.

In case of std goals (xor/ec are similar, but let's assume std for simplicity), with `REDUNDANCY_LEVEL = 0`, master will allow the write operation to proceed only if it can write one copy + 0 additional data copies.
With `REDUNDANCY_LEVEL = 2`, master will allow the write operation only if it can write one copy + **2** additional copies, which makes **3** copies total.

So, if any of your chunks resided on turned off chunkservers from `wks`, master would be able to modify only 2 copies of it, which is not enough to satisfy `REDUNDANCY_LEVEL = 2`.

tl;dr: The expected behaviour you describe in this issue is actually how `REDUNDANCY_LEVEL = 1` works. For std goals, `REDUNDANCY_LEVEL = 1` will force at least 2 copies to be modified.

**Comment by onlyjob** [comment]
Thank you very much for explaining that. It makes perfect sense now. :)

**Comment by onlyjob** [comment]
For replicated goals `REDUNDANCY_LEVEL = 1` seems to work but for erasure/xor goals the problem is just like I've described: consider goal `xor2: $xor2 {stor pool wks}` with multiple chunkservers in each label. I'm stopping 2 or more chunkservers in `wks` and modification of files with `xor2` goal becomes impossible... Seems like `REDUNDANCY_LEVEL = 1` does more harm than good...

**Comment by psarna** [comment]
Now that's because there's a potential conflict between `REDUNDANCY_LEVEL` and replication...
It's like that - let's assume that `REDUNDANCY_LEVEL` is 1, one of your chunks has xor2 goal and one of the chunkservers is off, so only 2 out of 3 parts of a chunk are available.

With `REDUNDANCY_LEVEL=1` you explicitly ask that every time something is written, it must be done on all data parts + at least 1 additional part. In order to immediately write to your chunk you would have to immediately replicate it and then write to all parts.

Now, what about replication limits? First of all, replication is by design independent of client operations, so triggering it synchronously on a write operation could create so many problems. If you touch 1000 chunks like that at the same time and write 1kB to each one, you would create a storm of replicating whole chunks, ending up with huge I/O burden, all at the same time. Operation timeouts pretty much guaranteed.
Also, replication has configurable limits in order to reduce I/O on chunkservers. Now, should we ignore them when writing to chunks like that? If not, then writing to such chunk should happen *after* a part is replicated, which can potentially take a long time, especially if all the limits are taken into account. Then, instead of "not enough chunkservers" error, you most likely get a timeout that doesn't really say much.

I don't think that `REDUNDANCY_LEVEL = 1` does more harm than good - chunks were read-only, fine, but at the same time they were exactly as safe as you wanted them to be, by manually setting redundancy level.

Anyway, if you or anyone else has an idea how to solve this issue better without affecting other LizardFS features (like replication and rebalancing), I'm open to suggestions.

**Comment by onlyjob** [comment]
`REDUNDANCY_LEVEL=1` makes cluster much less resilient to chunkserver failures. Of course it causes more harm and I've demonstrated that there are enough operational chunkservers to safely perform operation.

As for how... In case when all three chunks are available to re-write a file we read/access any two chunks then (re-)write three chunks. How is that different when one chunk is not available?? We still need to read two chunks and write three. One unavailable chunk should be queued for deletion anyway (if that can not be done then older revisions of chunks can be periodically scrapped to avoid bloating deletion queue with chunks that may never become available).
Why on earth chunk that is being modified can not be immediately replicated to the available chunkservers? Like when chunk is missing? Just treat offline/unavailable chunks as missing that's all.

Modifying many chunks at the same time is a different situation and timeouts are beyond the point...

**Comment by psarna** [comment]
Ok, I see the source of confusion now... the whole point is that when one of your ec parts is missing, it's not immediately replicated. Not at all. That's why we have redundancy level, endangered chunks priority, etc. Replication is **independent** of I/O operations. Chunks can be put in priority queue, yes, but nothing, ever, since LizardFS creation, was replicated immediately on write. That's the design of it and implementing immediate whole-chunk replication every time undergoal chunk is touched would create a ton of other issues.

Also, replication is 64MB chunk-wise, while I/O operations are 64kB block-wise. That's another reason why it's not trivial at all to combine replication+writes.

**Comment by guestisp** [comment]
@psarna just a little bit OT: so, chunk replication during I/O (i'm not talking about healing and similiar), but client1 is writing a huge file to LizardFS, is indipendent from any replication settings? In other words, even if I set a very very very very low replication setting, I/O writes are still replicated immediatly.

replication settings only affect what was already stored on the cluster, not what is being wrote "right now", right ? So, even by setting a replication at 1 chunk per year, **current I/O** is still replicated to all chunkservers (based on goal setting) immediatly and real-time, but, in case of failure (where replication/healing should start), what was already wrote (and thus need to be healed), is replicated 1 chunk per year. Right ?

**Comment by psarna** [comment]
Yes, if you write to goal 3, all copies are **written** before operation succeeds. When you lose one copy, it will be restored during **replication**. Writing is just **writing** to 3 copies during one operation, not writing to one and **replicating** immediately to another. These are different code patterns and have different rules.

**Comment by guestisp** [comment]
@psarna thanks for confirmation. Perfect.

**Comment by 4Dolio** [comment]
So then regarding https://github.com/lizardfs/lizardfs/issues/605 the client attempts the fulfill all goal satisfying writes for all goal encoding methods?  I use X (simple) and I believe XOR and EC also must write the entire goal as the client without resorting to replication, under normal circumstances with all CS targets working.  Replication of course would kick in when there is a missing CS member and goals have become incomplete.

Given that you have enough CS to satisfy goal copy spread plus some number of missing CS such that after replication you can still satisfy the minimum needed number of goal copies.  What would happen when you have not enough CS to hold each needed goal part?  I suspect a single CS is prevented from holding multiple pieces?  Perhaps this is causing problems or confusion and is impacted by REDUNDANCY_LEVEL settings also.

As I understand, If we used the simple goal of 3 as an example.  we must have 3 CS to satisfy the goal but this leaves no overhead.  Using REDUNDANCY_LEVEL=2(max lost before endangered) here would force 2 to be written, more would be considered endangered.  If only 1 CS were available it would be an endangered write and thus result in a write failure.  REDUNDANCY_LEVEL=1 would still succeed with only one CS available?

https://github.com/lizardfs/lizardfs/blob/master/src/data/mfsmaster.cfg.in#L266
> Minimum number of required redundant chunk parts that can be lost before chunk becomes endangered (Default: 0)

XOR2 Needs 3, but may survive with 2 (data+data) or either (data+parity). Not sure what REDUNDANCY_LEVEL does here XOR is always endangered when it loses REDUNDANCY_LEVEL=0 parts. XOR with REDUNDANCY_LEVEL=1 would still be undergoal, but not endangered I guess. If you used REDUNDANCY_LEVEL=1 then losing one XOR would be endangered and fail?

sigh, I think I lost myself, not sure if any of that makes any sends.

**Comment by psarna** [comment]
@4Dolio 
Before `REDUNDANCY_LEVEL` times, if you had just 1 chunkserver, LizardFS would allow you to write data, even if your goal = 15. In other words, creating undergoal files was allowed if there was not enough chunkservers in your installation.

With `REDUNDANCY_LEVEL` you can prevent this, relate to my reply in post number 4 for details about standard goals.

With ec/xor it's also fairly simple: with `REDUNDANCY_LEVEL = X` your write will only succeed if there's enough chunkservers to create **all** data parts and **X** parity parts. With `REDUNDANCY_LEVEL = 0` it's possible to write even if only data parts would be created (thus, creating endangered chunks).

**Comment by 4Dolio** [comment]
Ty @psarna. All clear.

**Comment by onlyjob** [comment]
Thanks for insightful explanations, @psarna. 

IMHO operation should succeed if there are enough chunkservers up to comply with goal _despite some offline chunks on unavailable/unreachable chunkservers_. If replication has to be forced prior to operation then so be it -- it will be slow but at least I/O will continue instead of failing right away.

Resilience to temporary unreachable/unavailable chunkservers is important.

It defeats the purpose to have many chunkservers for redundancy if operations are stopping in situation when 80% of chunkservers in one label are available to accommodate 1 replica of three (1 of 5 chunkservers is down).

**Comment by psarna** [comment]
We'll discuss this issue internally. I get that the way it works now is not perfect, but there are many technical barriers to simply change it. This topic is really close to two core functionalities of LizardFS, I/O mechanism and replication, and changing anything related to these two should be very careful.

Let this issue remain open, and if I (or anyone) have further comments or patches, they will be placed here.
